capsules/src/net/tlv.rs

```diff
@@ -0,0 +1,1324 @@
+pub enum NetworkDataTlvType {
+    Prefix = 1,
+    CommissioningData = 4,
+    Service = 5,
+    NotPresent,
+}
+
+impl From<u8> for NetworkDataTlvType {
+    fn from(tlv_type: u8) -> Self {
+        match tlv_type {
+            1 => NetworkDataTlvType::Prefix,
+            4 => NetworkDataTlvType::CommissioningData,
+            5 => NetworkDataTlvType::Service,
+            _ => NetworkDataTlvType::NotPresent,
+        }
+    }
+}
+
+impl<'a, 'b> From<&'a NetworkDataTlv<'b>> for NetworkDataTlvType {
+    fn from(network_data_tlv: &'a NetworkDataTlv<'b>) -> Self {
+        match *network_data_tlv {
+            NetworkDataTlv::Prefix { .. } => NetworkDataTlvType::Prefix,
+            NetworkDataTlv::CommissioningData { .. } => NetworkDataTlvType::CommissioningData,
+            NetworkDataTlv::Service { .. } => NetworkDataTlvType::Service,
+        }
+    }
+}
+
+pub enum PrefixSubTlv<'a> {
+    HasRoute(&'a [u8]),
+    BorderRouter(&'a [u8]),
+    SixLoWpanId {
+        context_id_compress: bool,
+        context_id: u8,
+        context_length: u8,
+    },
+}
+
+impl<'a> PrefixSubTlv<'a> {
+    pub fn encode(&self, buf: &mut [u8], stable: bool) -> SResult {
+        match *self {
+            PrefixSubTlv::HasRoute(ref r_border_router_16s) => {
+                let value_width = r_border_router_16s.len();
+                self.encode_tl(buf, value_width, stable);
+                let offset = enc_consume!(buf, TL_WIDTH; encode_bytes, r_border_router_16s);
+                stream_done!(offset)
+            }
+            PrefixSubTlv::BorderRouter(ref p_border_router_16s) => {
+                let value_width = p_border_router_16s.len();
+                self.encode_tl(buf, value_width, stable);
+                let offset = enc_consume!(buf, TL_WIDTH; encode_bytes, p_border_router_16s);
+                stream_done!(offset)
+            }
+            PrefixSubTlv::SixLoWpanId { context_id_compress, context_id, context_length } => {
+                let value_width = mem::size_of::<u8>() + mem::size_of::<u8>();
+                self.encode_tl(buf, value_width, stable);
+                let compress_bit = if context_id_compress { 1u8 } else { 0u8 };
+                let first_byte = (compress_bit << 4) | (context_id & 0b1111);
+                let mut offset = enc_consume!(buf, TL_WIDTH; encode_u8, first_byte);
+                offset = enc_consume!(buf, offset; encode_u8, context_length);
+                stream_done!(offset)
+            }
+        }
+    }
+
+    fn encode_tl(&self, buf: &mut [u8], value_width: usize, stable: bool) -> SResult {
+        stream_len_cond!(buf, TL_WIDTH + value_width);
+        let stable_bit = if stable { 1u8 } else { 0u8 };
+        buf[0] = (PrefixSubTlvType::from(self) as u8) << 1 | stable_bit;
+        buf[1] = value_width as u8;
+        stream_done!(2)
+    }
+
+    // Returns PrefixSubTlv and true if stable, false otherwise.
+    pub fn decode(buf: &[u8]) -> SResult<Option<(PrefixSubTlv, bool)>> {
```

Comment:
Fixed: Removed Option from all the return types. I was using it to return None in case the TLV type is not detected. Updated code handles it better, per a suggestion from Hubert.

The bool is to indicate whether the TLV data is 'stable' -- a status that's outlined in the spec.